packages/react-reconciler/src/ReactFiberTracingMarkerComponent.new.js

```diff
@@ -100,31 +101,77 @@ export function processTransitionCallbacks(
 // tracing marker can be logged as complete
 // This code lives separate from the ReactFiberTransition code because
 // we push and pop on the tracing marker, not the suspense boundary
-const tracingMarkerStack: StackCursor<Array<Fiber> | null> = createCursor(null);
+const markerInstanceStack: StackCursor<Array<TracingMarkerInstance> | null> = createCursor(
+  null,
+);
 
-export function pushTracingMarker(workInProgress: Fiber): void {
+export function pushRootMarkerInstance(workInProgress: Fiber): void {
   if (enableTransitionTracing) {
-    if (tracingMarkerStack.current === null) {
-      push(tracingMarkerStack, [workInProgress], workInProgress);
+    // On the root, every transition gets mapped to it's own map of
+    // suspense boundaries. The transition is marked as complete when
+    // the suspense boundaries map is empty. We do this because every
+    // transition completes at different times and depends on different
+    // suspense boundaries to complete. We store all the transitions
+    // along with its map of suspense boundaries in the root incomplete
+    // transitions map. Each entry in this map functions like a tracing
+    // marker does, so we can push it onto the marker instance stack
+    const transitions = getWorkInProgressTransitions();
+    const root = workInProgress.stateNode;
+
+    if (transitions !== null) {
+      transitions.forEach(transition => {
+        // We need to create a new map here because we only have access to the
+        // object instance in the commit phase
+        root.incompleteTransitions.set(transition, {
```

Comment:
Ah it shouldn't because we only push the root once and we only push the transition when it has just been added but I'll add a check here just to make sure!